pkg/metric/writer_prometheus.go

```diff
@@ -0,0 +1,221 @@
+package metric
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"github.com/justtrackio/gosoline/pkg/cfg"
+	"github.com/justtrackio/gosoline/pkg/clock"
+	"github.com/justtrackio/gosoline/pkg/log"
+	"github.com/justtrackio/gosoline/pkg/mdl"
+	gosoPrometheus "github.com/justtrackio/gosoline/pkg/metric/prometheus"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+const (
+	UnitPromCounter   StandardUnit = "prom-counter"
+	UnitPromGauge     StandardUnit = "prom-gauge"
+	UnitPromHistogram StandardUnit = "prom-histogram"
+	UnitPromSummary   StandardUnit = "prom-summary"
+)
+
+type promMetricFactory func(*promDatum) prometheus.Metric
+type promMetricPersister func(metric prometheus.Metric, data *promDatum)
+
+type promSettings struct {
+	// MetricLimit is used to avoid having metrics for which the name is programmatically generated (or have large number
+	// of possible dimensions) which could lead in a memory leak.
+	MetricLimit int64 `cfg:"metric_limit" default:"10000"`
+}
+
+type promDatum struct {
+	*Datum
+	namespace string
+}
+
+type promWriter struct {
+	logger      log.Logger
+	clock       clock.Clock
+	promMetrics sync.Map
+	registry    *prometheus.Registry
+	namespace   string
+	metricLimit int64
+	metrics     *int64
+}
+
+func NewPromWriter(ctx context.Context, config cfg.Config, logger log.Logger) (*promWriter, error) {
+	settings := &promSettings{}
+	config.UnmarshalKey("metric.prometheus", settings)
+
+	appId := cfg.GetAppIdFromConfig(config)
+	namespace := fmt.Sprintf("%s:%s:%s", appId.Project, appId.Environment, appId.Family)
```

Comment:
So the namespace is on family level? How do you distinguish between applications? I would have expected the app name here as well.

Reply:
added the application too, but now anyone could have it's own namespace naming strategy for prometheus by using the `metric.WithPromNSNamingStrategy()` func